src/index.js

```diff
@@ -1,4 +1,168 @@
+
+class IPLDResolver {
+  constructor (blockService) {
+    // nicola will love this!
+    if (!blockService) {
+      const repo = new IPFSRepo('in-memory', { stores: MemoryStore })
+      blockService = new BlockService(repo)
+    }
+
+    this.bs = blockService
+    this.resolvers = {}
+
+    // Support by default dag-pb and dag-cbor
+    this.support(dagPB.resolver.multicodec, dagPB.resolver, dagPB.util)
+    this.support(dagCBOR.resolver.multicodec, dagCBOR.resolver, dagCBOR.util)
+  }
+
+  // Adds support for an IPLD format
+  // default ones are dag-pb and dag-cbor
+  support (multicodec, resolver, util) {
+    this.resolvers[multicodec] = {
+      resolver: resolver,
+      util: util
+    }
+  }
+
+  resolve (cid, path, callback) {
+    if (path === '/') {
+      return this.get(cid, callback)
+    }
+
+    let value
+
+    until(
+      () => {
+        if (!path || path === '' || path === '/') {
+          return true
+        } else {
+          // continue traversing
+          if (value) {
+            cid = new CID(value['/'])
+          }
+          return false
+        }
+      },
+      (cb) => {
+        // get block
+        // use local resolver
+        // update path value
+        this.bs.get(cid, (err, block) => {
+          if (err) {
+            return cb(err)
+          }
+          const result = this.resolvers[cid.codec].resolver.resolve(block, path)
+          value = result.value
+          path = result.remainderPath
+          cb()
+        })
+      },
+      (err, results) => {
+        if (err) {
+          return callback(err)
+        }
+        return callback(null, value)
+      }
+    )
+  }
+
+  // Node operations (get and retrieve nodes, not values)
+
+  put (nodeAndCID, callback) {
+    callback = callback || noop
+    pull(
+      pull.values([nodeAndCID]),
+      this.putStream(callback)
+    )
+  }
+
+  putStream (callback) {
+    callback = callback || noop
+
+    return pull(
+      pull.map((nodeAndCID) => {
+        const cid = nodeAndCID.cid
+        const r = this.resolvers[cid.codec]
+        return {
+          block: new Block(r.util.serialize(nodeAndCID.node)),
+          cid: cid
+        }
+      }),
+      this.bs.putStream(),
+      pull.onEnd(callback)
+    )
+  }
+
+  get (cid, callback) {
+    pull(
+      this.getStream(cid),
+      pull.collect((err, res) => {
+        if (err) {
+          return callback(err)
+        }
+        callback(null, res[0])
+      })
+    )
+  }
+
+  getStream (cid) {
+    return pull(
+      this.bs.getStream(cid),
+      pull.map((block) => {
+        const r = this.resolvers[cid.codec]
+        if (r) {
+          return r.util.deserialize(block.data)
+        } else { // multicodec unknown, send back raw data
+          return block.data
+        }
+      })
+    )
+  }
+
+  // TODO: Consider if we still want this (fact: no one is using it
+  // right now and this will be just an IPLD selector anyway
+  getRecursive (cid, callback) {
```

Comment:
What we talked was `.tree` method inside an IPLD format. This `getRecursive` is an old method that comes from the previous "DAGService", which was never used in practice.